lib/maths/CBayesianOptimisation.cc

```diff
@@ -372,6 +398,268 @@ double CBayesianOptimisation::kernel(const TVector& a, const TVector& x, const T
+                if (core::CStringUtils::stringToType(traverser.value(), newValue) == false) {
+                    LOG_ERROR(<< "Error restoring vector: " << traverser.value());
+                    return false;
+                }
+                newVector.emplace_back(newValue);
+            }
+        } while (traverser.next());
+        vector = std::move(newVector);
+        return true;
+    };
+    return restorationFunction;
+}
+
+std::function<bool(core::CStateRestoreTraverser&)>
+CBayesianOptimisation::restoreVector(const std::string& tag,
+                                     CBayesianOptimisation::TVector& vector) {
+    auto restorationFunction = [&tag, &vector](core::CStateRestoreTraverser& traverser) {
+        double newValue{0.0};
+        CBayesianOptimisation::TDoubleVec newVector;
+        do {
+            if (traverser.name() == tag) {
+                if (core::CStringUtils::stringToType(traverser.value(), newValue) == false) {
+                    LOG_ERROR(<< "Error restoring vector: " << traverser.value());
+                    return false;
+                }
+                newVector.emplace_back(newValue);
+            }
+        } while (traverser.next());
+        CBayesianOptimisation::TVector newTVector(newVector.size());
+        for (std::size_t i = 0u; i < newVector.size(); ++i) {
+            newTVector(i) = newVector[i];
+        }
+        vector = std::move(newTVector);
+        return true;
+    };
+    return restorationFunction;
+}
+
+std::function<bool(core::CStateRestoreTraverser&)>
+CBayesianOptimisation::restoreParameterValuePair(TVector& parameters, double& functionValue) {
+    auto restoreParameterValuePair = [&parameters, &functionValue,
+                                      this](core::CStateRestoreTraverser& traverser) {
+        do {
+            if (traverser.name() == FUNCTION_PARAMETERS_TAG) {
+                if (traverser.traverseSubLevel(this->restoreVector(
+                        FUNCTION_PARAMETER_TAG, parameters)) == false) {
+                    return false;
+                }
+            } else if (traverser.name() == FUNCTION_VALUE_TAG) {
+                if (core::CStringUtils::stringToType(traverser.value(), functionValue) == false) {
+                    LOG_ERROR(<< "Error restoring function value: " << traverser.value());
+                    return false;
+                }
+            } else {
+                LOG_ERROR(<< "Error restoring function mean values. Unexpected tag: "
+                          << traverser.name());
+                return false;
+            }
+        } while (traverser.next());
+        return true;
+    };
+    return restoreParameterValuePair;
+}
+
+bool CBayesianOptimisation::restoreFunctionMeanValues(CBayesianOptimisation::TVectorDoublePrVec& functionMeanValues,
+                                                      core::CStateRestoreTraverser& traverser) {
+    if (traverser.hasSubLevel()) {
```

Comment:
I wouldn't bother with this (more on rationale in [this](TODO) comment).

Reply:
Here I need this check to make sure that restoring optimizer without any function values works correctly. Do you see any other solution?

Reply:
I think if this is written/read using core::CPersistUtils it handles empty collections by adding a "dummy" value to the state document (which is the alternative route).

This comment preceded the observation that that functionality could be used instead.

Reply:
Ok, I changed it to `core::CPersistUtils` now the methods become obsolete.